app/browser/windows.js

```diff
@@ -62,51 +61,44 @@ const updateWindow = (windowId) => {
   }
 }
 
+const siteMatchesTab = (site, tab) => {
+  const matchesLocation = getLocationIfPDF(tab.get('url')) === site.get('location')
+  const matchesPartition = (tab.get('partitionNumber') || 0) === (site.get('partitionNumber') || 0)
+  return matchesLocation && matchesPartition
+}
+
 const updatePinnedTabs = (win) => {
   if (win.webContents.browserWindowOptions.disposition === 'new-popup') {
     return
   }
-
   const appStore = require('../../js/stores/appStore')
   const state = appStore.getState()
   const windowId = win.id
   const pinnedSites = pinnedSitesState.getSites(state).map(site => pinnedSitesUtil.getPinnedSiteProps(site))
   const pinnedTabs = getPinnedTabsByWindowId(state, windowId)
-
-  pinnedSites.filter((site) =>
-    pinnedTabs.find((tab) =>
-      getLocationIfPDF(tab.get('url')) === site.get('location') &&
-      (tab.get('partitionNumber') || 0) === (site.get('partitionNumber') || 0))).forEach((site) => {
-        win.__alreadyPinnedSites = win.__alreadyPinnedSites.add(site)
+  // sites are instructions of what should be pinned
+  // tabs are sites our window already has pinned
+  const tabsWithSites = new Set()
+  for (const site of pinnedSites.values()) {
+    const existingPinnedTab = pinnedTabs.find(tab => siteMatchesTab(site, tab))
+    if (existingPinnedTab) {
+      tabsWithSites.add(existingPinnedTab)
+    } else {
+      appActions.createTabRequested({
+        url: site.get('location'),
+        partitionNumber: site.get('partitionNumber'),
+        pinned: true,
+        active: false,
+        windowId
       })
-
-  const sitesToAdd = pinnedSites.filter((site) =>
-    !win.__alreadyPinnedSites.find((pinned) => pinned.equals(site)))
-
-  sitesToAdd.forEach((site) => {
-    win.__alreadyPinnedSites = win.__alreadyPinnedSites.add(site)
-    appActions.createTabRequested({
-      url: site.get('location'),
-      partitionNumber: site.get('partitionNumber'),
-      pinned: true,
-      active: false,
-      windowId
-    })
-  })
-
-  const sitesToClose = win.__alreadyPinnedSites.filter((pinned) =>
-    !pinnedSites.find((site) => pinned.equals(site)))
-
-  sitesToClose
-    .forEach((site) => {
-      const tab = pinnedTabs.find((tab) =>
-        tab.get('url') === site.get('location') &&
-        (tab.get('partitionNumber') || 0) === (site.get('partitionNumber') || 0))
-      if (tab) {
-        appActions.tabCloseRequested(tab.get('tabId'), true)
-      }
-      win.__alreadyPinnedSites = win.__alreadyPinnedSites.remove(site)
-    })
+    }
+  }
+  for (const tab of pinnedTabs.values()) {
+    const shouldBePinned = tabsWithSites.has(tab)
+    if (!shouldBePinned) {
+      appActions.tabCloseRequested(tab.get('tabId'), true)
```

Comment:
In what case will require us to manually close tab?

Reply:
@darkdh this happens when you have > 1 window open, you then unpin a tab in window 1. This code calls upon window 2+ to close the previously-pinned tab. Whether or not this is what we want to happen I guess is beyond the scope of this bug fix, since that is the previous functionality. I myself have noticed some edge-cases that could end up with losing user webview state in a tab, but I'll log a separate issue for that...